cranelift/frontend/src/frontend.rs

```diff
@@ -704,45 +708,36 @@ impl<'a> FunctionBuilder<'a> {
-        let (safepoints, max_vals_in_stack_map_by_log2_size) =
-            self.find_live_stack_map_values_at_each_safepoint();
-
-        // Create the stack slots to spill them into.
-        let stack_slots = self.create_safepoint_slots(max_vals_in_stack_map_by_log2_size);
+        // Find all the GC values that are live across each safepoint and add
+        // stack map entries for them.
+        let stack_slots = self.find_live_stack_map_values_at_each_safepoint();
 
         // Insert spills to our new stack slots before each safepoint
         // instruction.
-        self.insert_safepoint_spills_to_stack_slots(safepoints, stack_slots);
+        self.insert_safepoint_spills_and_reloads(&stack_slots);
     }
 
     /// Find the live GC references for each safepoint instruction in this
     /// function.
     ///
-    /// Returns a pair of
-    ///
-    /// 1. A map from each safepoint instruction to the set of GC references
-    ///    that are live across it
-    ///
-    /// 2. The maximum number of values we need to store in a stack map at the
-    ///    same time, bucketed by their type's size. This array is indexed by
-    ///    the log2 of the type's size.
+    /// Returns a map from each safepoint instruction to the set of GC
+    /// references that are live across it
     fn find_live_stack_map_values_at_each_safepoint(
         &mut self,
-    ) -> (
-        crate::HashMap<ir::Inst, SmallVec<[ir::Value; 4]>>,
-        [usize; LOG2_SIZE_CAPACITY],
-    ) {
-        // A map from each safepoint to the set of GC references that are live
-        // across it.
-        let mut safepoints: crate::HashMap<ir::Inst, SmallVec<[ir::Value; 4]>> =
-            crate::HashMap::new();
-
-        // The maximum number of values we need to store in a stack map at the
-        // same time, bucketed by their type's size. This array is indexed by
-        // the log2 of the type's size. We do not support recording values whose
-        // size is greater than 16 in stack maps.
-        let mut max_vals_in_stack_map_by_log2_size = [0; LOG2_SIZE_CAPACITY];
+    ) -> crate::HashMap<ir::Value, ir::StackSlot> {
+        // A mapping from each needs-stack-map value that is live across some
+        // safepoint to the stack slot that it resides within. Note that if a
+        // needs-stack-map value is never live across a safepoint, then we won't
+        // ever add it to this map, it can remain in a virtual register for the
+        // duration of its lifetime, and we won't replace all its uses with
+        // reloads and all that stuff.
+        let mut stack_slots: crate::HashMap<ir::Value, ir::StackSlot> = Default::default();
+
+        // A map from size/align to free stack slots that are not being used
+        // anymore. This allows us to reuse stack slots across multiple values
+        // helps cut down on the ultimate size of our stack frames.
+        let mut free_stack_slots: crate::HashMap<u32, SmallVec<[ir::StackSlot; 4]>> =
```

Comment:
If it's not too bad to do, could we have an enum for stack-slot size rather than a bare `u32`? Something like `enum StackSlotSize { Slot4, Slot8 }` and perhaps `.bytes()` and `.align()` methods as needed; and a `TryFrom<ir::Type>` impl maybe too...

(To be clear, I'm suggesting this as a refactor wherever slot-sizes appear, the hashmap key here just prompted the thought)

Reply:
We would actually need quite a few slot sizes, since we have types like `f64x8` in CLIF. I'm not fully convinced that we would actually benefit from such an `enum`, and even if we do want to go down that route, I'd want to do it in a separate PR.

I did waffle back and forth between a full hash map here vs an array indexed by `log2(size)` like the old approach did all over the place. I opted to just go for the simplest, most-obvious implementation approach and figured we can come back and optimize if it is ever a perf bottleneck, but I suspect it really isn't.

So given all that: are you okay with keeping this as-is for now at least?

Reply:
Sure, it doesn't need to happen as part of this PR. (The reason for `TryFrom<ir::Type>` rather than `From` was exactly to address this -- I don't think we need to define a slot size for every type, only for types that the `cranelift-frontend` user puts into stackmaps -- and in general I find bare integers pretty error-prone; but happy to discuss further!)

Reply:
Did this in https://github.com/bytecodealliance/wasmtime/pull/8991